capsules/extra/src/i2c_nmea.rs

```diff
@@ -0,0 +1,137 @@
+    i2c: &'a I,
+    client: OptionalCell<&'a dyn NmeaClient>,
+}
+
+impl<'a, I: I2CDevice> I2cNmea<'a, I> {
+    pub fn new(i2c: &'a I, i2c_buffer: &'static mut [u8]) -> Self {
+        I2cNmea {
+            sentance_buffer: TakeCell::empty(),
+            i2c_buffer: TakeCell::new(i2c_buffer),
+            nmea_offset: Cell::new(0),
+            i2c,
+            client: OptionalCell::empty(),
+        }
+    }
+}
+
+impl<'a, I: I2CDevice> NmeaDriver<'a> for I2cNmea<'a, I> {
+    fn set_client(&self, client: &'a dyn NmeaClient) {
+        self.client.set(client);
+    }
+
+    fn read_sentance(
+        &self,
+        buffer: &'static mut [u8],
+    ) -> Result<(), (ErrorCode, &'static mut [u8])> {
+        self.sentance_buffer.replace(buffer);
+
+        self.i2c
+            .read(self.i2c_buffer.take().unwrap(), I2C_BUFFER_LEN)
+            .unwrap();
+
+        Ok(())
+    }
+}
+
+impl<'a, I: I2CDevice> I2CClient for I2cNmea<'a, I> {
+    fn command_complete(&self, buffer: &'static mut [u8], status: Result<(), i2c::Error>) {
+        let nmea_buf = self.sentance_buffer.take().unwrap();
+
+        if status.is_err() {
+            self.client.map(|call| {
+                call.callback(nmea_buf, 0, status);
+            });
+
+            self.i2c_buffer.replace(buffer);
+            return;
+        }
+
+        let string = str::from_utf8(buffer).unwrap();
+        let mut nmea_offset = self.nmea_offset.get();
+
+        if let Some(location) = string.find('$') {
+            // This is the start of a new sentance
+
+            if (nmea_offset + location) >= nmea_buf.len() {
+                // We will overflow our buffer, just drop the packet and try again
+                nmea_offset = 0;
+            }
+
+            // First copy the remainder of the sentance to the buffer
+            nmea_buf[nmea_offset..(nmea_offset + location)]
+                .copy_from_slice(string[0..location].as_bytes());
+            nmea_offset += location;
+
+            // Now parse the sentance
+            let sentance = str::from_utf8(&nmea_buf[0..nmea_offset]).unwrap();
+
+            if sentance.starts_with('$') {
+                // At this point we have a sentance with a `$` at the start.
+                // We report it back to the caller.
+                // We loose the rest of the data we just read though
+                self.client.map(|call| {
+                    call.callback(nmea_buf, nmea_offset, Ok(()));
+                });
+
+                self.i2c_buffer.replace(buffer);
+                nmea_offset = 0;
+
+                self.nmea_offset.set(nmea_offset);
+                return;
+            } else {
+                // The sentance didn't start with `$`. This usually occurs
+                // if we start reading mid-sentance. So we just try again and
+                // get the next one.
+                nmea_offset = 0;
+
+                // Copy the rest of the data
+                let size = I2C_BUFFER_LEN - location;
```

Comment:
This object should just use a fixed length buffer rather than having this constant which could not match the actual i2c buffer.